tests/server/test_desired_state_versions.py

```diff
@@ -0,0 +1,323 @@
+    await env.insert()
+
+    cm_timestamps = []
+    for i in range(1, 11):
+        cm_timestamps.append(datetime.datetime.strptime(f"2021-12-06T11:{i}:00.0", "%Y-%m-%dT%H:%M:%S.%f"))
+
+    # Add multiple versions of model
+    # 1: skipped_candidate, 2,3: retired, 4,5,6: skipped_candidate, 7: active, 8,9: candidate
+    for i in range(1, 10):
+        cm = data.ConfigurationModel(
+            environment=env.id,
+            version=i,
+            date=cm_timestamps[i],
+            total=1,
+            released=i in {2, 3, 7},
+            version_info={"export_metadata": {"message": "Recompile model because state transition", "type": "lsm_export"}}
+            if i % 2
+            else {
+                "export_metadata": {"message": "Recompile model because one or more parameters were updated", "type": "param"}
+            },
+        )
+        await cm.insert()
+    env2 = data.Environment(name="dev-test2", project=project.id, repo_url="", repo_branch="")
+    await env2.insert()
+    cm = data.ConfigurationModel(
+        environment=env2.id,
+        version=11,
+        date=datetime.datetime.now(),
+        total=1,
+        released=True,
+        version_info={},
+    )
+    await cm.insert()
+
+    env3 = data.Environment(name="dev-test3", project=project.id, repo_url="", repo_branch="")
+    await env3.insert()
+    cm = data.ConfigurationModel(
+        environment=env3.id,
+        version=7,
+        date=datetime.datetime.now(),
+        total=1,
+        released=False,
+        version_info={},
+    )
+    await cm.insert()
+
+    env4 = data.Environment(name="dev-test4", project=project.id, repo_url="", repo_branch="")
+    await env4.insert()
+    environments = {
+        "multiple_versions": env.id,
+        "single_released_version": env2.id,
+        "no_released_version": env3.id,
+        "no_versions": env4.id,
+    }
+
+    yield environments, cm_timestamps
+
+
+@pytest.mark.asyncio
+async def test_filter_versions(
+    server, client, environments_with_versions: Tuple[Dict[str, uuid.UUID], List[datetime.datetime]]
+):
+    """ Test querying desired state versions."""
+    environments, cm_times = environments_with_versions
+    env = environments["multiple_versions"]
+
+    result = await client.list_desired_state_versions(env)
+    assert result.code == 200
+    assert len(result.result["data"]) == 9
+
+    result = await client.list_desired_state_versions(env, filter={"version": ["ge:3", "le:5"]})
+    assert result.code == 200
+    assert len(result.result["data"]) == 3
+
+    result = await client.list_desired_state_versions(env, filter={"version": ["gt:10"]})
+    assert result.code == 200
+    assert len(result.result["data"]) == 0
+
+    result = await client.list_desired_state_versions(
+        env, filter={"date": [f"le:{cm_times[4].astimezone(datetime.timezone.utc)}"]}
+    )
+    assert result.code == 200
+    assert len(result.result["data"]) == 4
```

Comment:
Shouldn't this return 5 items, since the constraint is `le`?

Reply:
The indexing of the timestamps was off by one, with `timestamps[0]` being left unused, and `timestamps[1]` referring to the cm with version number 1 (but index 0). I realize that this can lead to confusion, so now the indexes match

Reply:
I see.